### DISCUSSION TITLE
Cache use for registered kafka listeners

### DISCUSSION THREAD

**Discussion by lkozien** [spring-projects/spring-integration#8872] [Q&A]
Hi, 

I am using `org.springframework.integration.channel.DirectChannel` with `org.springframework.kafka.listener.ConcurrentMessageListenerContainer` as Kafka consumer for my project. 

I would like to use @Cacheable method in service used by message execution. 

I found out that in `org.springframework.context.support.AbstractApplicationContext#refresh()` listeners are registered (and start to consume messages) before `org.springframework.context.support.AbstractApplicationContext#finishBeanFactoryInitialization(beanFactory)` which initialize Cache (`org.springframework.cache.interceptor.CacheAspectSupport`).

Is there any other way to use cache from the beginning of message consumption that to disable registration at autostart? For now cache is initialized and loaded after few messages consumed.

Best regards,
Lukasz

**Comment by artembilan** [comment]
Consider to use `CacheRequestHandlerAdvice` instead of `Cacheable` on your service activator definition: https://docs.spring.io/spring-integration/reference/handler-advice/classes.html#cache-advice.

It is also sounds suspicious that Kafka consumer has started to emit messages so early.
According to the `AbstractApplicationContext` logic:
```
				// Instantiate all remaining (non-lazy-init) singletons.
				finishBeanFactoryInitialization(beanFactory);

				// Last step: publish corresponding event.
				finishRefresh();
```
The `start()` on the `ConcurrentMessageListenerContainer` is called exactly from that `finishRefresh()` :
```
		// Propagate refresh to lifecycle processor first.
		getLifecycleProcessor().onRefresh();
```
